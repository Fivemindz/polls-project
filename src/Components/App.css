.dash-body {
  height: 1000px;
  display: grid;
  grid-template-rows: 200px auto;
  grid-template-columns: repeat(7, 1fr);
  grid-template-areas:
    " ban ban ban ban ban ban ban"
    "nv mn mn mn mn mn mn"
    ;
}
.container {
  margin: 10px 200px 5px 200px;
  padding: 10px;
}
.dash-loading {
  grid-area: mn;
  border: 1px solid lightgray;
  display: flex;
  align-items: center;
  justify-content: center;
}