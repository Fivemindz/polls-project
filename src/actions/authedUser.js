export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const RECEIVE_USER = "RECEIVE_USER";

export function setAuthedUser(user) {
  return {
    type: SET_AUTHED_USER,
    user,
  };
}

export function receiveUser(user) {
  return {
    type: RECEIVE_USER,
    user,
  };
}
