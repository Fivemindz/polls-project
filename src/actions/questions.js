import { saveQuestion, saveQuestionAnswer } from "../utils/api";
import { showLoading, hideLoading } from "react-redux-loading-bar";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const SAVE_QUESTION = "SAVE_QUESTIONS";
export const SAVE_QUESTION_ANSWER = "SAVE_QUESTION_ANSWER";

function addQuestion(question) {
  return {
    type: SAVE_QUESTION,
    question,
  };
}

function addQuestionAnswer(answer) {
  return {
    type: SAVE_QUESTION_ANSWER,
    answer,
  };
}

export function handleAddQuestion(optionOneText, optionTwoText) {
  return (dispatch, getState) => {
    const { authedUser } = getState();

    dispatch(showLoading());

    return saveQuestion({
      optionOneText,
      optionTwoText,
      author: authedUser,
    })
      .then((question) => {
        dispatch(addQuestion(question));
      })
      .then(() => dispatch(hideLoading()));
  };
}

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function handleAddQuestionAnswer(qid, answer) {
  return (dispatch, getState) => {
    const { authedUser } = getState();

    dispatch(showLoading());

    return saveQuestionAnswer({
      authedUser: authedUser,
      qid,
      answer,
    })
      .then((answer) => dispatch(addQuestionAnswer(answer)))
      .then(() => dispatch(hideLoading()));
  };
}
