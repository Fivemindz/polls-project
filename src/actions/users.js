export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_QUESTION_ANSWER = "ADD_QUESTION_ANSWER";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}
