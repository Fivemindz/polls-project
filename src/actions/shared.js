import { getInitialData, loadUsers } from "../utils/api";
import { receiveUsers } from "./users";
import { receiveQuestions } from "./questions";
import { setAuthedUser } from "./authedUser";
import { showLoading, hideLoading } from "react-redux-loading-bar";

export function handleInitialData(user) {
  return async (dispatch) => {
    dispatch(showLoading());
    return await getInitialData().then(({ users, questions }) => {
      dispatch(receiveUsers(users));
      dispatch(receiveQuestions(questions));
      dispatch(setAuthedUser(user));
      dispatch(hideLoading());
    });
  };
}

export function getUsers() {
  return async (dispatch) => {
    dispatch(showLoading());
    return loadUsers().then(({ users }) => {
      dispatch(receiveUsers(users));
      dispatch(hideLoading());
    });
  };
}
