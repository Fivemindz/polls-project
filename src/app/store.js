import { configureStore } from "@reduxjs/toolkit";
import { authedUser as authedUserReducer } from "../reducers/authedUser";
import { questions as questionsReducer } from "../reducers/questions";
import { users as usersReducer } from "../reducers/users";
import logger from "../middleware/logger";
import thunk from "redux-thunk";

export const store = configureStore({
  reducer: {
    authedUser: authedUserReducer,
    questions: questionsReducer,
    users: usersReducer,
  },
  middleware: [thunk],
});
