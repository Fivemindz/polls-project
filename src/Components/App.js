import { Fragment, useEffect, useState } from "react";
import { connect } from "react-redux";
import LoadingBar from "react-redux-loading-bar";
import Login from "./Login";
import { Routes, Route } from "react-router-dom";
import Banner from "./Banner";
import Nav from "./Nav";
import Dash from "./Dash";
import Stats from "./Stats";
import NewQuestion from "./NewQuestion";
import "./App.css";
import { getUsers } from "../actions/shared";
import QuestionPage from "./QuestionPage";
import Loading from "./Loading";
import PageNotFound from "./PageNotFound";

const App = (props) => {
  const { loading, dispatch, loadingBar } = props;
  const [usersLoaded, setUsersLoaded] = useState(false);

  useEffect(() => {
    const handleLoadUsers = async () => {
      await dispatch(getUsers());
      setUsersLoaded(true);
    };

    if (usersLoaded === false) {
      handleLoadUsers();
    }

    return () => {
      setUsersLoaded(true);
    };
  });

  return (
    <Fragment>
      <LoadingBar />
      <div className="container" data-testid="home-page">
        {loading === true ? (
          <Login />
        ) : (
          <div className="dash-body">
            <Banner />
            <Nav />
            {loadingBar ? (
              <div className="dash-loading">
                <Loading />
              </div>
            ) : (
              <Routes>
                <Route path="/" exact element={<Dash />} />
                <Route path="/stats" element={<Stats />} />
                <Route path="/newquestion" element={<NewQuestion />} />
                <Route path="/question/:id" element={<QuestionPage />} />
                <Route path="/notfound" element={<PageNotFound />} />
              </Routes>
            )}
          </div>
        )}
      </div>
    </Fragment>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  loading: authedUser === null,
});

export default connect(mapStateToProps)(App);
