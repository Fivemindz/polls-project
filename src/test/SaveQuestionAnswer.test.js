import { fireEvent, render, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "./store";
import { BrowserRouter as Router } from "react-router-dom";
import { handleInitialData } from "../actions/shared";
import App from "../Components/App";

describe("Save Question Workflow", () => {
  beforeAll(async () => {
    await store.dispatch(handleInitialData("tylermcginnis"));
  });

  it("Snapshot: Renders QuestionPage after clicking on question from dashboard", () => {
    const component = render(
      <Provider store={store}>
        <Router>
          <App />
        </Router>
      </Provider>
    );

    const link = component.getAllByText("sarahedo");
    fireEvent.click(link[0]);
    expect(component).toMatchSnapshot();
  });

  it("Loads stats for question after answer is submitted", async () => {
    const component = render(
      <Provider store={store}>
        <Router>
          <App />
        </Router>
      </Provider>
    );

    const link = component.getAllByText("tylermcginnis");
    fireEvent.click(link[0]);

    const answerCheckbox = component.getByTestId("answer-checkbox");
    const btn = component.getByTestId("answer-btn");
    fireEvent.click(answerCheckbox);
    fireEvent.click(btn);

    expect(component).toMatchSnapshot();
  });

  it("Displays message when user trys to submit without answer", () => {
    const component = render(
      <Provider store={store}>
        <Router>
          <App />
        </Router>
      </Provider>
    );

    const link = component.getAllByText("tylermcginnis");
    fireEvent.click(link[0]);

    const btn = component.getByTestId("answer-btn");
    fireEvent.click(btn);

    waitFor(() => {
      const selectOptionMessage = component.getByTestId(
        "select-option-message"
      );
      expect(selectOptionMessage).toBeInTheDocument();
    });
  });
});
