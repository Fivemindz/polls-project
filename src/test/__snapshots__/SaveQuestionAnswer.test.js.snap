// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Save Question Workflow Loads stats for question after answer is submitted 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="direction: ltr;"
      >
        <div
          class=""
          style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
        />
        <div
          style="display: table; clear: both;"
        />
      </div>
      <div
        class="container"
        data-testid="home-page"
      >
        <div
          class="dash-body"
        >
          <div
            class="user-banner"
          >
            <div
              class="user-info"
            >
              <img
                alt="avatar"
                class="avatar"
                src="./avatars/tyler.jpg"
              />
              <div
                class="user-name"
              >
                 
                tylermcginnis
              </div>
            </div>
            <div
              class="banner-spacer"
            />
            <div
              class="btn-container"
            >
              <button
                class="sign-out-btn"
              >
                Sign Out
              </button>
            </div>
          </div>
          <div
            class="nav"
          >
            <ul>
              <a
                class="nav-link"
                href="/"
              >
                <li>
                  Dashboard
                </li>
              </a>
              <a
                class="nav-link"
                href="/stats"
              >
                <li>
                  Stats
                </li>
              </a>
              <a
                class="nav-link"
                href="/newquestion"
              >
                <li>
                  New Poll
                </li>
              </a>
            </ul>
          </div>
          <div
            class="view-question"
          >
            <h2>
              View Poll by 
              sarahedo
               
            </h2>
            <img
              alt="avatar"
              class="author-avatar"
              src="../avatars/sarah.jpg"
            />
            <form
              class="view-question-form"
            >
              <h3>
                Would You Rather?
              </h3>
              <div
                class="question-area"
              >
                <label>
                  Answer 1
                </label>
                <div
                  class="answer-select"
                >
                  <input
                    data-testid="answer-checkbox"
                    id="optionOne"
                    type="checkbox"
                  />
                  <textarea
                    id="optionOneText"
                    readonly=""
                  >
                    conduct a release retrospective 1 week after a release
                  </textarea>
                </div>
                <label>
                  Answer 2
                </label>
                <div
                  class="answer-select"
                >
                  <input
                    id="optionTwo"
                    type="checkbox"
                  />
                  <textarea
                    id="optionTwoText"
                    readonly=""
                  >
                    conduct release retrospectives quarterly
                  </textarea>
                </div>
              </div>
              <button
                data-testid="answer-btn"
                type="submit"
              >
                Submit Answer
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="direction: ltr;"
    >
      <div
        class=""
        style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
      />
      <div
        style="display: table; clear: both;"
      />
    </div>
    <div
      class="container"
      data-testid="home-page"
    >
      <div
        class="dash-body"
      >
        <div
          class="user-banner"
        >
          <div
            class="user-info"
          >
            <img
              alt="avatar"
              class="avatar"
              src="./avatars/tyler.jpg"
            />
            <div
              class="user-name"
            >
               
              tylermcginnis
            </div>
          </div>
          <div
            class="banner-spacer"
          />
          <div
            class="btn-container"
          >
            <button
              class="sign-out-btn"
            >
              Sign Out
            </button>
          </div>
        </div>
        <div
          class="nav"
        >
          <ul>
            <a
              class="nav-link"
              href="/"
            >
              <li>
                Dashboard
              </li>
            </a>
            <a
              class="nav-link"
              href="/stats"
            >
              <li>
                Stats
              </li>
            </a>
            <a
              class="nav-link"
              href="/newquestion"
            >
              <li>
                New Poll
              </li>
            </a>
          </ul>
        </div>
        <div
          class="view-question"
        >
          <h2>
            View Poll by 
            sarahedo
             
          </h2>
          <img
            alt="avatar"
            class="author-avatar"
            src="../avatars/sarah.jpg"
          />
          <form
            class="view-question-form"
          >
            <h3>
              Would You Rather?
            </h3>
            <div
              class="question-area"
            >
              <label>
                Answer 1
              </label>
              <div
                class="answer-select"
              >
                <input
                  data-testid="answer-checkbox"
                  id="optionOne"
                  type="checkbox"
                />
                <textarea
                  id="optionOneText"
                  readonly=""
                >
                  conduct a release retrospective 1 week after a release
                </textarea>
              </div>
              <label>
                Answer 2
              </label>
              <div
                class="answer-select"
              >
                <input
                  id="optionTwo"
                  type="checkbox"
                />
                <textarea
                  id="optionTwoText"
                  readonly=""
                >
                  conduct release retrospectives quarterly
                </textarea>
              </div>
            </div>
            <button
              data-testid="answer-btn"
              type="submit"
            >
              Submit Answer
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Save Question Workflow Snapshot: Renders QuestionPage after clicking on question from dashboard 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="direction: ltr;"
      >
        <div
          class=""
          style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
        />
        <div
          style="display: table; clear: both;"
        />
      </div>
      <div
        class="container"
        data-testid="home-page"
      >
        <div
          class="dash-body"
        >
          <div
            class="user-banner"
          >
            <div
              class="user-info"
            >
              <img
                alt="avatar"
                class="avatar"
                src="./avatars/tyler.jpg"
              />
              <div
                class="user-name"
              >
                 
                tylermcginnis
              </div>
            </div>
            <div
              class="banner-spacer"
            />
            <div
              class="btn-container"
            >
              <button
                class="sign-out-btn"
              >
                Sign Out
              </button>
            </div>
          </div>
          <div
            class="nav"
          >
            <ul>
              <a
                class="nav-link"
                href="/"
              >
                <li>
                  Dashboard
                </li>
              </a>
              <a
                class="nav-link"
                href="/stats"
              >
                <li>
                  Stats
                </li>
              </a>
              <a
                class="nav-link"
                href="/newquestion"
              >
                <li>
                  New Poll
                </li>
              </a>
            </ul>
          </div>
          <div
            class="view-question"
          >
            <h2>
              View Poll by 
              sarahedo
               
            </h2>
            <img
              alt="avatar"
              class="author-avatar"
              src="../avatars/sarah.jpg"
            />
            <form
              class="view-question-form"
            >
              <h3>
                Would You Rather?
              </h3>
              <div
                class="question-area"
              >
                <label>
                  Answer 1
                </label>
                <div
                  class="answer-select"
                >
                  <input
                    data-testid="answer-checkbox"
                    id="optionOne"
                    type="checkbox"
                  />
                  <textarea
                    id="optionOneText"
                    readonly=""
                  >
                    conduct a release retrospective 1 week after a release
                  </textarea>
                </div>
                <label>
                  Answer 2
                </label>
                <div
                  class="answer-select"
                >
                  <input
                    id="optionTwo"
                    type="checkbox"
                  />
                  <textarea
                    id="optionTwoText"
                    readonly=""
                  >
                    conduct release retrospectives quarterly
                  </textarea>
                </div>
              </div>
              <button
                data-testid="answer-btn"
                type="submit"
              >
                Submit Answer
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="direction: ltr;"
    >
      <div
        class=""
        style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
      />
      <div
        style="display: table; clear: both;"
      />
    </div>
    <div
      class="container"
      data-testid="home-page"
    >
      <div
        class="dash-body"
      >
        <div
          class="user-banner"
        >
          <div
            class="user-info"
          >
            <img
              alt="avatar"
              class="avatar"
              src="./avatars/tyler.jpg"
            />
            <div
              class="user-name"
            >
               
              tylermcginnis
            </div>
          </div>
          <div
            class="banner-spacer"
          />
          <div
            class="btn-container"
          >
            <button
              class="sign-out-btn"
            >
              Sign Out
            </button>
          </div>
        </div>
        <div
          class="nav"
        >
          <ul>
            <a
              class="nav-link"
              href="/"
            >
              <li>
                Dashboard
              </li>
            </a>
            <a
              class="nav-link"
              href="/stats"
            >
              <li>
                Stats
              </li>
            </a>
            <a
              class="nav-link"
              href="/newquestion"
            >
              <li>
                New Poll
              </li>
            </a>
          </ul>
        </div>
        <div
          class="view-question"
        >
          <h2>
            View Poll by 
            sarahedo
             
          </h2>
          <img
            alt="avatar"
            class="author-avatar"
            src="../avatars/sarah.jpg"
          />
          <form
            class="view-question-form"
          >
            <h3>
              Would You Rather?
            </h3>
            <div
              class="question-area"
            >
              <label>
                Answer 1
              </label>
              <div
                class="answer-select"
              >
                <input
                  data-testid="answer-checkbox"
                  id="optionOne"
                  type="checkbox"
                />
                <textarea
                  id="optionOneText"
                  readonly=""
                >
                  conduct a release retrospective 1 week after a release
                </textarea>
              </div>
              <label>
                Answer 2
              </label>
              <div
                class="answer-select"
              >
                <input
                  id="optionTwo"
                  type="checkbox"
                />
                <textarea
                  id="optionTwoText"
                  readonly=""
                >
                  conduct release retrospectives quarterly
                </textarea>
              </div>
            </div>
            <button
              data-testid="answer-btn"
              type="submit"
            >
              Submit Answer
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
